### PR TITLE
Datetime compatibility to MySQL restored

### PR DESCRIPTION
**Pull request by annando**
newer versions of Oracle's MySQL hadn't accepted ```0000-00-00``` as datetime value anymore. We now have a constant ```NULL_DATE``` that indicates the lowest possible date value this now is defined as ```0001-01-01```. 

We should **always** use this constant in the future since it could always happen that another SQL server has different minimum datetime values. For example MS SQL has ```1753-01-01```as lowest datetime value.

To have no compatibility issues with old data in the tables we mustn't look for values different than the ```NULL_DATE```but lower or equal than this value.

Update: When reviewing, please have a look especially for the queries that replaces the "unequal 0000-00-00" statements. I hope that I haven't overlooked anything, but one never knows.

And: You will see some "0000-00-00" still in the code for the birthday functionality. This is no problem since the used fields are varchar fields.

### CODE REVIEW

**Comment by Hypolite** [comment]
Naive question: Why can't we just use `null` for null dates?

**Comment by annando** [comment]
At first I think that this wouldn't simplify the queries - and additionally we already had used the value ```0000-00-00``` as default value, so this would complicate the queries even more.

**Comment by Hypolite** [comment]
Oh yeah, it would definitely not simplify the query, we just wouldn't be using an absurd string around. Gotcha, commencing review now.

File: include/cron.php [hunk 1]
```diff
@@ -215,7 +215,7 @@ function cron_poll_contacts($argc, $argv) {
 			$xml = false;
 
 			if($manual_id)
```
Comment (Hypolite) [inline include/cron.php]: Standards: Can you please add braces to this condition?

File: boot.php [hunk 2]
```diff
@@ -441,6 +441,13 @@
...
 
+/**
+ * Lowest possible date time value
+ */
+
+define ('NULL_DATE', '0001-01-01 00:00:00');
+
+
 // Normally this constant is defined - but not if "pcntl" isn't installed
 if (!defined("SIGTERM"))
```
Comment (Hypolite) [inline boot.php]: Standards: Can you please add braces to this condition?

File: include/cron.php [hunk 3]
```diff
@@ -215,7 +215,7 @@ function cron_poll_contacts($argc, $argv) {
...
-				$contact['last-update'] = '0000-00-00 00:00:00';
+				$contact['last-update'] = NULL_DATE;
 
 			if(in_array($contact['network'], array(NETWORK_DFRN, NETWORK_ZOT, NETWORK_OSTATUS)))
```
Comment (Hypolite) [inline include/cron.php]: Standards: Can you please add spaces and braces to this condition?

File: include/poller.php [hunk 4]
```diff
@@ -366,8 +366,8 @@ function poller_kill_stale_workers() {
 
 	foreach($r AS $pid)
```
Comment (Hypolite) [inline include/poller.php]: Standards: Can you please add braces to this loop?

File: include/poller.php [hunk 5]
```diff
@@ -366,8 +366,8 @@ function poller_kill_stale_workers() {
...
 		if (!posix_kill($pid["pid"], 0))
```
Comment (Hypolite) [inline include/poller.php]: Standards: Can you please add braces to this condition?

File: include/poller.php [hunk 6]
```diff
@@ -553,20 +554,24 @@ function poller_worker_process() {
 	if (poller_passing_slow($highest_priority)) {
 		// Are there waiting processes with a higher priority than the currently highest?
 		$r = q("SELECT * FROM `workerqueue`
-				WHERE `executed` = '0000-00-00 00:00:00' AND `priority` < %d
-				ORDER BY `priority`, `created` LIMIT 1", dbesc($highest_priority));
+				WHERE `executed` <= '%s' AND `priority` < %d
+				ORDER BY `priority`, `created` LIMIT 1",
+				dbesc(NULL_DATE),
+				intval($highest_priority));
 		if (dbm::is_result($r))
```
Comment (Hypolite) [inline include/poller.php]: Standards: Can you please add braces to this condition?

File: include/poller.php [hunk 7]
```diff
@@ -553,20 +554,24 @@ function poller_worker_process() {
...
-				WHERE `executed` = '0000-00-00 00:00:00' AND `priority` > %d
-				ORDER BY `priority`, `created` LIMIT 1", dbesc($highest_priority));
+				WHERE `executed` <= '%s' AND `priority` > %d
+				ORDER BY `priority`, `created` LIMIT 1",
+				dbesc(NULL_DATE),
+				intval($highest_priority));
 	}
 
 	// If there is no result (or we shouldn't pass lower processes) we check without priority limit
 	if (($highest_priority == 0) OR !dbm::is_result($r))
```
Comment (Hypolite) [inline include/poller.php]: Standards: Can you please add braces to this condition?

File: include/socgraph.php [hunk 8]
```diff
@@ -242,7 +242,7 @@ function poco_check($profile_url, $name, $network, $profile_photo, $about, $loca
 		if (($network == "") AND ($x[0]["network"] != NETWORK_STATUSNET))
 			$network = $x[0]["network"];
 
-		if ($updated == "0000-00-00 00:00:00")
+		if ($updated <= NULL_DATE)
```
Comment (Hypolite) [inline include/socgraph.php]: Standards: Can you please add braces to this condition?

File: include/socgraph.php [hunk 9]
```diff
@@ -242,7 +242,7 @@ function poco_check($profile_url, $name, $network, $profile_photo, $about, $loca
 		if (($network == "") AND ($x[0]["network"] != NETWORK_STATUSNET))
```
Comment (Hypolite) [inline include/socgraph.php]: Standards: Can you please add braces to this condition?

File: include/socgraph.php [hunk 10]
```diff
@@ -465,7 +465,7 @@ function poco_last_updated($profile, $force = false) {
 	$gcontacts = q("SELECT * FROM `gcontact` WHERE `nurl` = '%s'",
 			dbesc(normalise_link($profile)));
 
-	if ($gcontacts[0]["created"] == "0000-00-00 00:00:00")
+	if ($gcontacts[0]["created"] <= NULL_DATE)
```
Comment (Hypolite) [inline include/socgraph.php]: Standards: Can you please add braces to this condition?

File: include/socgraph.php [hunk 11]
```diff
@@ -666,7 +666,7 @@ function poco_last_updated($profile, $force = false) {
 	// Maybe there aren't any entries. Then check if it is a valid feed
 	if ($last_updated == "")
 		if ($xpath->query('/atom:feed')->length > 0)
```
Comment (Hypolite) [inline include/socgraph.php]: Standards: Can you please merge and add braces to those conditions?

File: include/socgraph.php [hunk 12]
```diff
@@ -946,7 +946,7 @@ function poco_check_server($server_url, $network = "", $force = false) {
 	$servers = q("SELECT * FROM `gserver` WHERE `nurl` = '%s'", dbesc(normalise_link($server_url)));
 	if (dbm::is_result($servers)) {
 
-		if ($servers[0]["created"] == "0000-00-00 00:00:00")
+		if ($servers[0]["created"] <= NULL_DATE)
```
Comment (Hypolite) [inline include/socgraph.php]: Standards: Can you please add braces to this condition?

File: mod/contacts.php [hunk 13]
```diff
@@ -558,7 +558,7 @@ function contacts_content(App $a) {
 		// tabs
 		$tab_str = contacts_tab($a, $contact_id, 2);
 
-		$lost_contact = (($contact['archive'] && $contact['term-date'] != '0000-00-00 00:00:00' && $contact['term-date'] < datetime_convert('','','now')) ? t('Communications lost with this contact!') : '');
+		$lost_contact = (($contact['archive'] && $contact['term-date'] > NULL_DATE && $contact['term-date'] < datetime_convert('','','now')) ? t('Communications lost with this contact!') : '');
 
 		if ($contact['network'] == NETWORK_FEED)
```
Comment (Hypolite) [inline mod/contacts.php]: Standards: Can you please add braces to this condition?

File: mod/contacts.php [hunk 14]
```diff
@@ -534,11 +534,11 @@ function contacts_content(App $a) {
 
 		$insecure = t('Private communications are not available for this contact.');
 
-		$last_update = (($contact['last-update'] == '0000-00-00 00:00:00')
+		$last_update = (($contact['last-update'] <= NULL_DATE)
 				? t('Never')
 				: datetime_convert('UTC',date_default_timezone_get(),$contact['last-update'],'D, j M Y, g:i A'));
 
-		if($contact['last-update'] !== '0000-00-00 00:00:00')
+		if ($contact['last-update'] > NULL_DATE)
```
Comment (Hypolite) [inline mod/contacts.php]: Standards: Can you please add braces to this condition?

File: mod/profiles.php [hunk 15]
```diff
@@ -234,7 +234,7 @@ function profiles_post(App $a) {
 		$with = ((x($_POST,'with')) ? notags(trim($_POST['with'])) : '');
 
 		if(! strlen($howlong))
```
Comment (Hypolite) [inline mod/profiles.php]: Standards: Can you please add braces to this condition?

**Comment by annando** [comment]
Concerning the review please keep in mind that the more lines are changed for standards stuff, the more merge conflicts will be created.

**Comment by Hypolite** [comment]
I can take care of the merge conflicts if any..

File: include/cron.php [hunk 16]
```diff
@@ -214,11 +214,13 @@ function cron_poll_contacts($argc, $argv) {
...
+			if ($manual_id) {
 				$contact['last-update'] = NULL_DATE;
+			}
 
-			if(in_array($contact['network'], array(NETWORK_DFRN, NETWORK_ZOT, NETWORK_OSTATUS)))
+			if (in_array($contact['network'], array(NETWORK_DFRN, NETWORK_ZOT, NETWORK_OSTATUS))) {
 				$contact['priority'] = 2;
+			}
 
 			if($contact['subhub'] AND in_array($contact['network'], array(NETWORK_DFRN, NETWORK_ZOT, NETWORK_OSTATUS))) {
```
Comment (Hypolite) [inline include/cron.php]: Standards: Please add a space after `if`.

File: include/poller.php [hunk 17]
```diff
@@ -400,6 +400,7 @@ function poller_kill_stale_workers() {
 			} else
```
Comment (Hypolite) [inline include/poller.php]: Standards: Please add braces to this `else`.

File: include/socgraph.php [hunk 18]
```diff
@@ -664,10 +664,11 @@ function poco_last_updated($profile, $force = false) {
 	}
 
 	// Maybe there aren't any entries. Then check if it is a valid feed
-	if ($last_updated == "")
-		if ($xpath->query('/atom:feed')->length > 0)
+	if ($last_updated == "") {
+		if ($xpath->query('/atom:feed')->length > 0) {
```
Comment (Hypolite) [inline include/socgraph.php]: Can you please merge those conditions to remove an indentation level?

**Comment by annando** [comment]
Let's see what you will find now.

**Pull request merged**